modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/trait/TraitUtils.java

```diff
@@ -123,16 +126,23 @@ else if (converter == RewindabilityTraitDef.INSTANCE)
     }
 
     /** */
-    @Nullable public static RelNode convertCollation(RelOptPlanner planner,
-        RelCollation toTrait, RelNode rel) {
+    @Nullable public static RelNode convertCollation(RelOptPlanner planner, RelCollation toTrait, RelNode rel) {
         RelCollation fromTrait = collation(rel);
 
         if (fromTrait.satisfies(toTrait))
             return rel;
 
         RelTraitSet traits = rel.getTraitSet().replace(toTrait);
 
-        return new IgniteSort(rel.getCluster(), traits, rel, toTrait);
+        Sort sortNode = null;
+
+        if (rel instanceof SortNode)
+            sortNode = (Sort)rel;
+        if (rel instanceof RelSubset && ((RelSubset)rel).getOriginal() instanceof Sort)
```

Comment:
Looks fragile. Can we implement this logic in `SortConverterRule`?

Reply:
Fixed